### PR TITLE
Fix LaTeX escaping in text nodes and add unicode conversion

### LINKED ISSUE #28: JuliaConsoleLexer produces wrong output for Array
The JuliaConsoleLexer adds some extra braces to the output part of the following code.

```
highlight(stdout,MIME("text/latex"),"""julia> [1,2,3]
       3-element Array{Int64,1}:
        1
        2
        3""",Highlights.Lexers.JuliaConsoleLexer)
```
Which gives this latex code:
```
\begin{lstlisting}
(*@\HLJLnB{julia> }@*)(*@\HLJLp{[}@*)(*@\HLJLni{1}@*)(*@\HLJLp{,}@*)(*@\HLJLni{2}@*)(*@\HLJLp{,}@*)(*@\HLJLni{3}@*)(*@\HLJLp{]}@*)
3-element Array{\{}Int64,1{\}}:
 1
 2
 3
\end{lstlisting}
```
image

### PR DESCRIPTION
**Pull request by sebastianpech**
Continuation of #29.
- Fix #28 
- Add unicode to latex conversion

### CODE REVIEW

**Comment by sebastianpech** [comment]
I have a running version that manages to parse (ie. no LaTeX error) 2417 chars out of the 2492 defined in `REPL.REPLCompletions.latex_symbols`. Some of the chars don't show up because of the selected font.
My current approach is like this:
1. Check if there is a mapping defined in this [database](http://milde.users.sourceforge.net/LUCR/Math/#database) (The db is licensed with LaTeX Project Public License, not sure if using only a subset (ie. manipulating the data) is allowed. However the database was already used similarly and I contacted the maintainer (no response yet).
2. If no mapping was found, try to parse the LaTeX-like code in  `REPL.REPLCompletions.latex_symbols`.

Some points open for discussion (@mortenpi, ...):
1. The above procedure relies on the user compiling the document with XeLaTeX or LuaLaTeX because I'm using the [unicode-math](https://ctan.org/pkg/unicode-math?lang=de) package.
2. For compiling I need a few extra packages that need to be added in the stylesheet: amssymb, wasysym, unicode-math, xfrac, bm

Is this something that is acceptable? The alternative would be to don't use [unicode-math](https://ctan.org/pkg/unicode-math?lang=de) and the additional packages. However the number of symbols that can be parsed would be reduced.

**Comment by KristofferC** [comment]
Personally, I think it is fine to require xelatex or lualatex. Good unicode support is just too important for Julia.

**Comment by sebastianpech** [comment]
This is a side-by-side comparison of the old output of Weave+Highlights and the new one. 
What I did in Weave was just removing the unicode escaping there and fully rely on Highlights.
In comparison to the escaping in weave now also
- inline fractures
- correct handling of math / not-math mode
- special formats like sans, tt, ...
- sub- and superscript
- special unicode chars without actual LaTeX representation 

I used the XITSMath font thats also documented in unicode-math
image

**Comment by sebastianpech** [comment]
Hmm, I just read the unicode-math documentation again and it seems that the package just converts LaTeX macros back to unicode, which makes the extra parsing useless as the outcome still depends on the selected font.

In my option there are the following options:

1. Don't escape unicode at all and leave handling to the user. This means the user must pick a font that supports the unicode characters (not latin modern, something like Xits) and this also means the user must compile with XeLaTeX or LuaLaTeX.
2. Only escape the characters that have a proper LaTeX math representation (not sure how many, but for sure considerably less. I can check that). This way, quite likely, a lot of commonly used characters are supported and the user is free to used whatever tex-engine and font he/she likes.

This could be controlled through a global variable. Considering that (2) might be sufficient for most users it could be the default. If someone wants (1) for example in Weave this could then simply be done by adding the font loading to the template file and setting a global switch in Highlights to pass through unicode characters unescaped.

What do you think about that?

**Comment by sebastianpech** [comment]
Updated list with extra 161 chars: [latex_unicode_map.pdf](https://github.com/JuliaDocs/Highlights.jl/files/3513310/latex_unicode_map.pdf)

**Comment by mortenpi** [comment]
Apologies, I forgot about this PR. What's the current state of this?

* Regarding additional dependencies -- sure, no problem, whatever is necessary to make it work well.
* If it's not too hard to do, having a font-independent fallback (with lower coverage) would be a good feature I think. But it really depends on what the dependents like Weave need.
* FWIW, I switched Documenter's PDF monospace font over to [DejaVu Sans Mono](https://dejavu-fonts.github.io/) not too long ago, to get better Unicode coverage. But there are still symbols that are not covered (e.g. `⊻`).

**Comment by sebastianpech** [comment]
TBH I think there is no proper solution to this without adding too many latex dependencies. As we don't have control over what latex packages the user has installed that's not an option. Doing no escaping and letting the user handle unicode chars is the safest bet I think. 

I'll check the commits on the weekend and just keep the stuff that's not directly related to unicode escaping (eg. escaping of double quotes for babel support)

**Comment by sebastianpech** [comment]
Changes:
- Initial problem with text mode, verbatim mode escaping
- Also escape "
- Remove textcomp and fontenc. The user must load fontspec for xelatex or textcomp + fontenc ... for pdflatex

**Comment by sebastianpech** [comment]
If using escapeinside there is a problem with multiline strings and comments because LaTeX removes single line breaks and leading spaces by default. To preserve them they must be explicitly placed (ie. \n -> \\newline and ' ' -> \\mbox{\\space})

File: src/Format.jl [hunk 1]
```diff
@@ -223,22 +222,34 @@ function escape(io::IO, ::MIME"text/html", str::AbstractString)
...
+        char === '#'   ? printe(io, charescape, "{\\#}") :
+        char === '&'   ? printe(io, charescape, "{\\&}") :
+        char === '\\'  ? printe(io, charescape, "{\\textbackslash}") :
+        char === '^'   ? printe(io, charescape, "{\\textasciicircum}") :
+        char === '_'   ? printe(io, charescape, "{\\_}") :
+        char === '{'   ? printe(io, charescape, "{\\{}") :
+        char === '}'   ? printe(io, charescape, "{\\}}") :
+        char === '~'   ? printe(io, charescape, "{\\textasciitilde}") :
+        char === '"'   ? printe(io, charescape, "\"{}") :
+        # Linebreaks within an escapeinside don't occour if not replaced by proper
```
Comment (mortenpi) [inline src/Format.jl]:
```suggestion
        # Linebreaks within an escapeinside do not occur if not replaced by proper
```

**Comment by mortenpi** [comment]
LGTM. For the user, this would be a seamless fix right (e.g. it doesn't rely on any new deps etc.)? Any tests that could be added to showcase the fixed escapes?

**Comment by sebastianpech** [comment]
I added some tests for the escape function, mainly to check if charescape true/false works as expected. About breaking something: The only change that could have an effect is the change in the stylesheet. However, if one already uses XeLaTeX that's not a problem (Weave defaults to it, so thats ok). For those using pdfLaTeX the removed lines need to be added to the preamble.

**Comment by sebastianpech** [comment]
Is this going to be merged someday maybe?

**Comment by mortenpi** [comment]
Sorry, I forgot about this. Thanks for bumping. LGTM, will merge and tag.

**Pull request merged**